File: python/render_sdk/experimental/object/e2e_helpers.py
"""Standalone CRUD check for object storage.

No pytest dependency — reusable by pollinators / health probes.
"""

import asyncio
import contextlib
import uuid

from render_sdk.client.errors import RenderError
from render_sdk.experimental.object.client import ObjectClient

RETRY_ATTEMPTS = 5
RETRY_INTERVAL_S = 1.0


async def object_storage_crud_check(
    client: ObjectClient, owner_id: str, region: str
) -> str:
    """Run a full PUT → LIST → GET → DELETE → verify-deletion cycle.

    Returns the key that was used (caller can use it for follow-up assertions).
    """
    key = f"e2e-test/{uuid.uuid4()}/crud-test.txt"
    content = b"hello from e2e"

    try:
        # --- PUT ---
        await client.put(
            owner_id=owner_id,
            region=region,
            key=key,
            data=content,
        )

        # --- LIST (with retry — eventual consistency) ---
        found = False
        for attempt in range(1, RETRY_ATTEMPTS + 1):
            list_response = await client.list(owner_id=owner_id, region=region)
            match = next((o for o in list_response.objects if o.key == key), None)
            if match is not None:
                if match.size != len(content):
                    raise AssertionError(
                        f"LIST size mismatch: expected {len(content)}, got {match.size}"
                    )
                found = True
                break
            if attempt < RETRY_ATTEMPTS:
                await asyncio.sleep(RETRY_INTERVAL_S)

        if not found:
            raise AssertionError(
                f"Object {key} not found in LIST after {RETRY_ATTEMPTS} attempts"
            )

        # --- GET ---
        obj = await client.get(owner_id=owner_id, region=region, key=key)
        if obj.data != content:
            raise AssertionError("GET data does not match uploaded content")
        if obj.size != len(content):
            raise AssertionError(
                f"GET size mismatch: expected {len(content)}, got {obj.size}"
            )
        # --- DELETE ---
        await client.delete(owner_id=owner_id, region=region, key=key)

        # --- Verify deletion ---
        try:
            await client.get(owner_id=owner_id, region=region, key=key)
            raise AssertionError("GET after DELETE should have raised")
        except RenderError:
            pass  # Expected — object was deleted

        return key
    except Exception:
        with contextlib.suppress(Exception):
            await client.delete(owner_id=owner_id, region=region, key=key)
        raise

File: typescript/src/experimental/object/e2e-helpers.ts
/**
 * Standalone CRUD check for object storage.
 *
 * No test-framework dependency — reusable by pollinators / health probes.
 */

import { randomUUID } from "node:crypto";
import { RenderError } from "../../errors.js";
import type { ObjectClient } from "./client.js";

const RETRY_ATTEMPTS = 5;
const RETRY_INTERVAL_MS = 1000;

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Run a full PUT → LIST → GET → DELETE → verify-deletion cycle.
 *
 * @returns The key that was used (caller can use it for follow-up assertions).
 */
export async function objectStorageCrudCheck(
  client: ObjectClient,
  ownerId: `tea-${string}`,
  region: string,
): Promise<string> {
  const key = `e2e-test/${randomUUID()}/crud-test.txt`;
  const content = Buffer.from("hello from e2e");

  try {
    // --- PUT ---
    await client.put({
      ownerId,
      region,
      key,
      data: content,
    });

    // --- LIST (with retry — eventual consistency) ---
    let found = false;
    for (let attempt = 1; attempt <= RETRY_ATTEMPTS; attempt++) {
      const listResponse = await client.list({ ownerId, region });
      const match = listResponse.objects.find((o) => o.key === key);
      if (match) {
        if (match.size !== content.byteLength) {
          throw new Error(`LIST size mismatch: expected ${content.byteLength}, got ${match.size}`);
        }
        // lastModified should be recent (within 5 minutes)
        const age = Date.now() - match.lastModified.getTime();
        if (age > 5 * 60 * 1000) {
          throw new Error(`LIST lastModified too old: ${match.lastModified.toISOString()}`);
        }
        found = true;
        break;
      }
      if (attempt < RETRY_ATTEMPTS) {
        await sleep(RETRY_INTERVAL_MS);
      }
    }
    if (!found) {
      throw new Error(`Object ${key} not found in LIST after ${RETRY_ATTEMPTS} attempts`);
    }

    // --- GET ---
    const obj = await client.get({ ownerId, region, key });
    if (!obj.data.equals(content)) {
      throw new Error("GET data does not match uploaded content");
    }
    if (obj.size !== content.byteLength) {
      throw new Error(`GET size mismatch: expected ${content.byteLength}, got ${obj.size}`);
    }
    // --- DELETE ---
    await client.delete({ ownerId, region, key });

    // --- Verify deletion ---
    try {
      await client.get({ ownerId, region, key });
      throw new Error("GET after DELETE should have thrown");
    } catch (err) {
      if (!(err instanceof RenderError)) {
        throw err;
      }
      // Expected — object was deleted
    }

    return key;
  } catch (err) {
    // Best-effort cleanup on failure
    try {
      await client.delete({ ownerId, region, key });
    } catch {
      // swallow cleanup errors
    }
    throw err;
  }
}
